add options to #initialize to allow configurable sanitizable content …

switzersc opened the pull request:
…types

I have apps that support media types like 'application/vnd.api+json' in the Content-Type header and I would like to sanitize that data. I added a simple options hash to the initialize method so that users can configure additional content types to be sanitized (or replace the sanitized content types altogether if they want to). Thoughts?

whitequark commented:
You broke CI. Also, I want tests for this.

switzersc commented:
Lol. I should've added that I couldn't get the tests to run locally, although now that I see the build I know why. Will update.

switzersc commented:
Fixed the build. Tests coming.

bf4 commented:
Alternatively, you could just document https://github.com/whitequark/rack-utf8_sanitizer/issues/12#issuecomment-46449510 https://github.com/whitequark/rack-utf8_sanitizer/commit/6f00635a1b1be571d2d41e84c41c08745e2c0d18

e.g. 

```ruby
Rack::UTF8Sanitizer::SANITIZABLE_CONTENT_TYPES.concat [
  'application/vnd.api+json'
]
```

whitequark commented:
Mutating a constant? That's gross. Frankly both the array and the strings inside should have been frozen anyway.

switzersc commented:
Thanks for both of your input! I hadn't seen that issue. I prefer configuring middleware behavior in the application.rb file over in an initializer, and point taken about mutating constants. I've updated the code in `initialize` to add the optional array to `SANITIZABLE_CONTENT_TYPES` instead of concating it, which mutates it, and I've added tests. Let me know if you have any other comments or suggestions :)

whitequark commented:
Looks good to me--would you also add a commit to freeze the default `SANITIZABLE_CONTENT_TYPES`? Something like `[…].map(&:freeze).freeze`.

switzersc commented:
Sure. Is it worth also freezing the other constants? `URI_FIELDS` and `URI_ENCODED_CONTENT_TYPES`

whitequark commented:
Yeah, all of them.

switzersc commented:
Updated. @bf4, just a heads up, this should break your workaround.

bf4 commented on lib/rack/utf8_sanitizer.rb:
```diff
@@ -7,8 +7,12 @@ module Rack
+    def initialize(app, options={})
       @app = app
+      @sanitizable_content_types = options[:sanitizable_content_types]
+      @sanitizable_content_types ||= SANITIZABLE_CONTENT_TYPES + (options[:additional_content_types] || [])
```
I think this should be a set operation, no?

whitequark replied:
It's not important in the context of this PR, but yes, `SANITIZABLE_CONTENT_TYPES` is semantically a set and there could be some minor performance benefit from switching to that as well.

bf4 commented on lib/rack/utf8_sanitizer.rb:
```diff
@@ -23,18 +27,18 @@ def call(env)
     URI_ENCODED_CONTENT_TYPES = %w(
       application/x-www-form-urlencoded
-    )
+    ).map(&:freeze).freeze
```
@whitequark what's the advantage of freezing constant-level strings? Just that it would blow up on attempt to change?

whitequark replied:
Yeah, it's good practice.
